src/fd.cc

```diff
@@ -250,10 +250,11 @@ void
 fd_note(int fd, const char *s)
 {
     fde *F = &fd_table[fd];
-    if (s)
-        xstrncpy(F->desc, s, FD_DESC_SZ);
+    safe_free(F->desc);
+    if (!s)
+        F->desc = xstrdup(s);
```

Comment:
IIRC, `desc` is only used for debugging/reporting. If possible, we should not add frequent (multiple times per transaction) memory allocations/deallocations for those purposes.

The old preallocated FD_DESC_SZ buffer approach was better than the PR version, but it can be improved further to avoid memory copies in most cases. If you want to improve it, consider changing `desc` type to SBuf while adjusting most fd_note() callers to supply statically allocated constant SBufs. SBuf was designed for this sort of thing.

Reply:
61 bytes is very much insufficient to contain even modern domains, let alone the URL it is supposed to be reporting.

Changing it to SBuf was the first thing I tried. It had two problems;
 * First that SBuf adds allocate/deallocate cycle same as this patch in addition to a secondary allocate/deallocate on each mgr report. You made it pretty clear in previous PR that touching those mgr reports even to fix typos is going to be vetoed unless the PR redesigns them wholesale and completely into absolutely perfect code.
 * Second it makes everything in Squid depend on libsbuf. Including libraries libsbuf itself depends on. So we get circular reference in linker dependency checking. Here again the mgr reports (for SBuf ones) are going to need a redesign to remove that dependency.

The next thing I looked at was using static strings the desc member just pointed at. This has issues again with the URL cases where the content has to be dynamically assigned. We could work around that with a paired ID+string design for v5+, but that is too much change for this fix which needs to go into v4 stables.

Reply:
NP: I'm not seeking to improve the desc handling, just to resolve the issue that several KB of URL string are being stuffed into 61B memory. Increasing FD_DESC_SZ is not the solution it appears to be as that would require many MB of RAM to be allocated and not used (most Squid already allocate over 4MB there and never use it).